### ISSUE TITLE
[Question] deserialize any type to string during runtime

### ISSUE THREAD

**Issue by lostiniceland** [sfackler/rust-postgres#858]
I am having trouble getting the deserialization via `row.get(0)` to work. My problem is, that get is generic for `T` and the conversion is applied at compile-time depending on the target, but in my case I just want to deserialize any value to String during runtime (the column-type is only known at runtime).

`thread 'main' panicked at 'error retrieving column 0: error deserializing column 0: cannot convert between the Rust type '&[u8]' and the Postgres type 'numeric'`

My use-case is that I can pass in arbitrary sql-queries from the outside and I just want to format the resultset as a ascii-table.

My last attempt was to try to get a `[u8]` deserialized and create a String from this.

```rust
impl<'set> From<Vec<postgres::Row>> for Table {
    fn from(item: Vec<postgres::Row>) -> Self {
        let columns = item
            .first()
            .map(|row| row.columns())
            .unwrap_or(&[])
            .iter()
            .map(|col| col.name().to_string())
            .collect();
        let rows = item
            .iter()
            .map(|row| {
                let range = (0..row.len());
                let mut table_row = Vec::with_capacity(range.len());
                // FIXME desierialization not working
                for index in range.step_by(1) {
                    let x: &[u8] = row.get::<usize, &[u8]>(index);
                    table_row.push(String::from_utf8_lossy(x).to_string());
                }
                table_row
            })
            .collect();
        Table::new(columns, rows)
    }
}
```
Here, `Table` is just my simple struct to save the column-headers and data as Strings.

I cant seem to find any API to just get a raw string from a row-value. Or do I have to write custom pattern-match or trait-impls for all kind of db-types which map to String?

**Comment by lostiniceland** [comment]
After digging into some source I've built a "reflective" get-function which matches on the column-type and converts accordingly

      ```
  /// The postgres-crate does not provide a default mapping to fallback to String for all
        /// types: row.get is generic and without a type assignment the FromSql-Trait cannot be inferred.
        /// This function matches over the current column-type and does a manual conversion
        fn reflective_get(row: &postgres::Row, index: usize) -> String {
            let column_type = row.columns().get(index).map(|c| c.type_().name()).unwrap();
            println!("----- {}", column_type);
            // see https://docs.rs/sqlx/0.4.0-beta.1/sqlx/postgres/types/index.html
            let value = match column_type {
                "bool" => {
                    let v: Option<bool> = row.get(index);
                    v.map(|v| v.to_string())
                }
                "varchar" | "char(n)" | "text" | "name" => {
                    let v: Option<String> = row.get(index);
                    v
                }
                // "char" => {
                //     let v: i8 = row.get(index);
                // }
                "int2" | "smallserial" | "smallint" => {
                    let v: Option<i16> = row.get(index);
                    v.map(|v| v.to_string())
                }
                "int" | "int4" | "serial" => {
                    let v: Option<i32> = row.get(index);
                    v.map(|v| v.to_string())
                }
                "int8" | "bigserial" | "bigint" => {
                    let v: Option<i64> = row.get(index);
                    v.map(|v| v.to_string())
                }
                "float4" | "real" => {
                    let v: Option<f32> = row.get(index);
                    v.map(|v| v.to_string())
                }
                "float8" | "double precision" => {
                    let v: Option<f64> = row.get(index);
                    v.map(|v| v.to_string())
                }
                "timestamp" | "timestamptz" => {
                    // with-chrono feature is needed for this
                    let v: Option<chrono::DateTime<Utc>> = row.get(index);
                    v.map(|v| v.to_string())
                }
                &_ => Some("CANNOT PARSE".to_string()),
            };
            value.unwrap_or("".to_string())
        }
```
There are still some edge-cases and I need to create a test-setup with a row containing all available postgres-types but I guess that is the correct way to handle this.

Maybe a get_as_string function would be nice to avoid this custom code.

**Comment by kevlarr** [comment]
@lostiniceland I actually want this exact same functionality; should this be reopened?

**Comment by lostiniceland** [comment]
@kevlarr not sure if there is a reason for this. I came to realize that all the rust-database-clients I currently use look strikingly similar: they all have a FromSql trait and have a generic `row.get(index)` function. For instance, in [tiberius](https://docs.rs/tiberius/0.4.5/tiberius/) you have to do a [similar matching](https://github.com/prisma/tiberius/issues/101#issuecomment-727295516) on the columntype to convert the data to String.
The only database-client which supports direct String conversion is the [oracle](https://crates.io/crates/oracle) crate, though this one is not a pure Rust driver because it relies on a C runtime client to be present on the system. I should admit that I currently only know those 3 clients for Rust (Oracle, Postgres, MSSQL).

So long story short: I dont know if there is a bigger reason why the a String conversion (for any db type) is not automatically implemented for pure Rust database drivers. Mabye  @sfackler could comment on that.

On the other side, this might be a nice opportunity to add something via PR.
